src/utils/helper_functions.py

```diff
@@ -22,17 +21,22 @@ def concatenate_inputs(
     # add all obs columns in the unintegrated data to integrated data
     # loc should order the obs based on obs_names
     
-    input_integrated.obs = input_unintegrated.obs.loc[input_integrated.obs_names]
+    adata = input_integrated.copy()
     
+    obs_reference_data = ad.concat([input_unintegrated, input_validation])
+    adata.obs = obs_reference_data.obs.loc[adata.obs_names]
+    
+    # oh man i end up having to do this because otherwise some metrics which does 
+    # pairwise comparison between samples from the same donor will give you error..
+    if adata.uns['method_id'] == 'perfect_integration':
```

Comment:
@LuLeom I ended up still having to do this as the sample names for the integrated data produced by the positive control is exactly the same as the one in validation. This will cause metrics that does sample-pairwise calculation (like EMD) to give out NaN. So in reality, for the controls, we can just return all obs and vars info and update this exception here and say if method_id is perfect_integration or no_integration then don't embed embed anything.